
## 中通面试总结

**面试过程主要设计以下问题：**
- jvm，内存模型，gc root,调优
- 线程池注意事项，异常处理
- 锁,分布式锁
- redis原生命令，持久化，高可用
- rpc框架实现原理,通讯协议，数据结构（zookeeper）
- ES
- mq,kafka
- 分布式系统：分布式事务，全链路追踪，限流(自实现)，分流（应用场景，mq削峰）
- 接口调用变慢排查
- 业务系统架构，业务量
- 数据库设计，优化方案
- 入职规划


以上问题中有部分回答比较欠缺，下面针对不熟悉知识点加以整理复习；


#### 1. gc root 怎么选择？存放在哪儿？

**GC Roots原理**

GC Roots基本思路就是通过一系列的称为“GC Roots”的对象作为起始点， 从这些节点开始向下搜索， 
搜索所走过的路径称为引用链（ Reference Chain），当一个对象到 GC Roots 没有任何引用链相连（ 用图论的话来 说，就是从GC Roots到这个对象不可达）时，
则证明此对象是不可用的。

**GC Roots对象**

常说的GC(Garbage Collector) Roots，特指的是垃圾收集器（Garbage Collector）的对象，GC会收集那些不是GC Roots且没有被GC Roots引用的对象。

一个对象可以属于多个root，GC Roots有以下几种：

- Class - 由系统类加载器(system class loader)加载的对象，这些类是不能够被回收的，他们可以以静态字段的方式保存持有其它对象。
  我们需要注意的一点就是，通过用户自定义的类加载器加载的类，除非相应的Java.lang.Class实例以其它的某种（或多种）方式成为roots，否则它们并不是roots，.
- Thread - 活着的线程
- Stack Local - Java方法的local变量或参数
- JNI Local - JNI方法的local变量或参数
- JNI Global - 全局JNI引用
- Monitor Used - 用于同步的监控对象
- Held by JVM - 用于JVM特殊目的由GC保留的对象，但实际上这个与JVM的实现是有关的。可能已知的一些类型是：系统类加载器、一些JVM知道的重要的异常类、一些用于处理异常的预分配对象以及一些自定义的类加载器等。然而，JVM并没有为这些对象提供其它的信息，因此需要去确定哪些是属于"JVM持有"的了。

#### 2. JVM 调优实战


#### 3. 阻塞队列的大小及设定？线程池出现异常是怎么处理？

```java
public static ExecutorService newFixedThreadPool(int nThreads) {
    return new ThreadPoolExecutor(nThreads, nThreads,
                                  0L, TimeUnit.MILLISECONDS,
                                  new LinkedBlockingQueue<Runnable>());
}

/**
 * 默认阻塞队列大小为 2^31-1，所以在使用时，必须
 */
public LinkedBlockingQueue() {
    this(Integer.MAX_VALUE);
}

/**
 * 必须设定阻塞队列大小
 */
public LinkedBlockingQueue(int capacity) {
    if (capacity <= 0) throw new IllegalArgumentException();
    this.capacity = capacity;
    last = head = new Node<E>(null);
}
```

**线程池异常怎么处理？**

在实际开发中，我们常常会用到线程池，但任务一旦提交到线程池之后，如果发生异常之后，怎么处理？ 
怎么获取到异常信息？而不是任务提交之后，消失的无影无踪。

要知道以上答案，先看下 线程池里面的线程发生异常之后会发生什么？

我们先通过工具类Executors创建一个简单的线程池，里面核心线程数为1

```java
  ExecutorService executorService=Executors.newFixedThreadPool(1);
  
   executorService.submit(()->{
            int i=1/0;
        });


  executorService.submit(()->{
        System.out.println("当线程池抛出异常后继续新的任务");
    });
```

上面我们创建了一个核心线程数和最大线程数都为1的线程池。然后往里面提交了两个任务。

其中一个任务 int i=1/0; 0不能作为除数，因此会抛出异常，java线程中 如果抛出未被捕获的异常，会导致线程终止。

该线程池中只有一个线程，如果终止之后，提交第二个任务会发生什么？

运行结果

![运行结果](https://img-blog.csdnimg.cn/20190906010418659.png)

答案是**第一个线程内部发生异常之后，没有任何异常信息出现，第二个任务正常执行。**

线程池有两种提交方式 excute和sumbit ，换成excute 看下

```java
  ExecutorService executorService=Executors.newFixedThreadPool(1);

        executorService.execute(()->{
            int i=1/0;
        });

        executorService.execute(()->{
            System.out.println("当线程池抛出异常后继续新的任务");
        });
```
![](https://img-blog.csdnimg.cn/20190906010627844.png)

第一个任务出现了异常栈信息，第二个任务正常执行。

线程池的两种不同的提交方式，会有不同的异常情形，但是不管怎样，线程内的任务抛出异常之后，线程池照样能正常运行。

**问题1: execute和submit有什么区别？为什么一个会抛出异常 一个不会？**

这个问题先放这，我们等下回头再看，这个问题非常重要。

那么当前最要紧的问题来了，我们要如何才能获取到线程池里面的任务抛出的异常？
```java
ExecutorService.execute(()->{
    try{
        int i=1/0;
    }catch (Exception ex){
        System.out.println(ex.getMessage());
    }
});
```

换成submit提交

```java
executorService.submit(()->{
    try{
        int i=1/0;
    }catch (Exception ex){
        System.out.println("sumbit提交"+ex.getMessage());
    }
});

executorService.submit(()->{
    System.out.println("当线程池抛出异常后继续新的任务");
});
```


可以看到 清晰易懂的捕获到了异常，可以知道我们的任务出现了问题，而不是消失的无影无踪。

**解析方案2: 每一个任务都加一个try-catch 实在是太麻烦了，而且代码也不好看，那么这样想的话，
可以用UncaughtExceptionHandler 这个类。**


UncaughtExceptionHandler 是Thread类一个内部类，也是一个函数式接口。

内部的uncaughtException是一个处理线程内发生的异常的方法，参数为线程对象t和异常对象e。

使用方式如下
```java
//创建线程对象 内部会抛出异常
Thread thread=new Thread(()->{
        int i=1/0;
    });

//设置该对象的默认异常处理器
thread.setDefaultUncaughtExceptionHandler((Thread t, Throwable e)->{
    System.out.println("exceptionHandler"+e.getMessage());
 });

//启动线程
thread.start();
```

运行结果
![运行结果](https://img-blog.csdnimg.cn/20190906012936817.png)

相反，如果我们不设置UncaughtExceptionHandler ，那么就是

        Thread thread=new Thread(()->{
            int i=1/0;
        });

        thread.start();

![](https://img-blog.csdnimg.cn/2019090601303220.png)

直接抛出异常。

因此 Thread的UncaughtExceptionHandler类能帮我们捕获异常并处理， 那么在线程池里面生效吗？

**试试excute提交**
```java

  ExecutorService executorService=Executors.newFixedThreadPool(1);

        Thread thread=new Thread(()->{
            int i=1/0;
        });


        thread.setDefaultUncaughtExceptionHandler((Thread t, Throwable e)->{
            System.out.println("exceptionHandler"+e.getMessage());
         });

   

      executorService.execute(thread);
```
![](https://img-blog.csdnimg.cn/20190906013221617.png)

在excute提交方式里面是生效的。


那么因此，如果我们不想在每个线程的任务里面都加try-catch的话，可以自己实现的一个线程池，重写它的线程工厂方法，在创建线程的时候，都赋予UncaughtExceptionHandler处理器对象。

具体代码如下

```java
        //1.实现一个自己的线程池工厂
        ThreadFactory factory = (Runnable r) -> {
            //创建一个线程
            Thread t = new Thread(r);
            //给创建的线程设置UncaughtExceptionHandler对象 里面实现异常的默认逻辑
            t.setDefaultUncaughtExceptionHandler((Thread thread1, Throwable e) -> {
                System.out.println("线程工厂设置的exceptionHandler" + e.getMessage());
            });
            return t;
        };

        //2.创建一个自己定义的线程池，使用自己定义的线程工厂
        ExecutorService service = new ThreadPoolExecutor(1, 1, 0, TimeUnit.MILLISECONDS,new LinkedBlockingQueue(10),factory);

        //3.提交任务
        service.execute(()->{
            int i=1/0;
        });
```

测试发现，即使我们不用在thread里面try-catch 也能拿到异常信息了。


#### 4. redis原生命令，持久化，高可用



#### 5. 分布式锁实现原理？

[请参考：分布式锁](../part6-framework/6-2-3%20分布式锁.md)

#### 6. 如何自己实现一个rpc框架？



#### 7. 分布式事务

[请参考：分布式事务](../part6-framework/6-2-4%20分布式事务.md)

#### 8. 分库分表方案








